src/Attributes.jl

```diff
@@ -0,0 +1,210 @@
+            AbstractAlgebra._get_attributes(::Type{$expr}) = $attr
+         end)
+      end
+   end
+   error("attributes can only be attached to mutable structs")
+end
+
+"""
+    _get_attributes(G::T) where T
+
+Return the dictionary storing the attributes for `G` if available, otherwise
+return `nothing`. If type `T` does not support attribute storage, an exception
+is thrown.
+
+This is used to implement `has_attribute` and `get_attribute`.
+Alternate attribute storage solutions need only provide custom methods
+for `_get_attributes` and `_get_attributes!`.
+"""
+function _get_attributes(G::T) where T
+   isstructtype(T) && ismutable(T) && hasfield(T, :other) || error("attributes storage not supported for type $T")
+   return isdefined(G, :other) ? G.other : nothing
+end
+
+"""
+    _get_attributes!(G::T) where T
+
+Return the dictionary storing the attributes for `G`, creating it if necessary.
+If type `T` does not support attribute storage, an exception is thrown.
+
+This is used to implement `set_attribute!` and `get_attribute!`.
+Alternate attribute storage solutions need only provide custom methods
+for `_get_attributes` and `_get_attributes!`.
+directly.
+"""
+function _get_attributes!(G::T) where T
+   isstructtype(T) && ismutable(T) && hasfield(T, :other) || error("attributes storage not supported for type $T")
+   if !isdefined(G, :other)
+      G.other = Dict{Symbol, Any}()
+   end
+   return G.other
+end
+
+"""
+    has_attribute(G::Any, attr::Symbol)
+
+Return a boolean indicating whether `G` has a value stored for the attribute `attr`.
+"""
+function has_attribute(G::Any, attr::Symbol)
+   D = _get_attributes(G)
+   return D isa Dict && haskey(D, attr)
+end
+
+"""
+    get_attribute(G::Any, attr::Symbol, default::Any = nothing)
+
+Return the value stored for the attribute `attr`, or if no value has been set,
+return `default`.
+"""
+function get_attribute(G::Any, attr::Symbol, default::Any = nothing)
+   D = _get_attributes(G)
+   D isa Dict && return get(D, attr, default)
+   return default
+end
+
+"""
+    get_attribute!(f::Function, G::Any, attr::Symbol)
+
+Return the value stored for the attribute `attr` of `G, or if no value has been set,
+store `key => f()` and return `f()`.
+"""
+function get_attribute!(f::Function, G::Any, attr::Symbol)
+   D = _get_attributes!(G)
+   return Base.get!(f, D, attr)
+end
+
+"""
+    get_attribute!(G::Any, attr::Symbol, default::Any)
+
+Return the value stored for the attribute `attr` of `G`, or if no value has been set,
+store `key => default`, and return `default`.
+"""
+function get_attribute!(G::Any, attr::Symbol, default::Any)
+   D = _get_attributes!(G)
+   return Base.get!(D, attr, default)
+end
+
+"""
+    set_attribute!(G::Any, data::Pair{Symbol, <:Any}...)
+
+Attach the given sequence of `key=>value` pairs as attributes of `G`.
+"""
+function set_attribute!(G::Any, data::Pair{Symbol, <:Any}...)
+   D = _get_attributes!(G)
+   for d in data
+      push!(D, d)
```

Comment:
Is `D` a dictionary?

Reply:
Yes. This code is copied from the old `set_special`